Switch to using the `sphinx-autoapi` extension for auto building the API reference. It better maintained and produces nicer formatted docs.

cdbf1 opened the pull request:
The sphinx auto-summary, auto-doc and apidoc extensions have more recently been improved upon by the single extension `sphinx-autoapi` which I think is simpler and build nicer looking documentation. 

This PR updates our docs to use `sphinx-autoapi`

vtomole reviewed (commented):
After running `./checks/build_docs.py` locally I get


```
WARNING: Cannot resolve import of cirq_superstaq.ops.qubit_gates.AQTICCX in cirq_superstaq.ops
WARNING: Cannot resolve import of cirq_superstaq.ops.qubit_gates.AQTITOFFOLI in cirq_superstaq.ops
WARNING: Cannot resolve import of cirq_superstaq.ops.qubit_gates.CR in cirq_superstaq.ops
WARNING: Cannot resolve import of cirq_superstaq.ops.qubit_gates.ZX in cirq_superstaq.ops
WARNING: Cannot resolve import of cirq_superstaq.ops.AQTICCX in cirq_superstaq
WARNING: Cannot resolve import of cirq_superstaq.ops.AQTITOFFOLI in cirq_superstaq
WARNING: Cannot resolve import of cirq_superstaq.ops.CR in cirq_superstaq
WARNING: Cannot resolve import of cirq_superstaq.ops.ZX in cirq_superstaq
```

but the docs do indeed build. I think it's fine to merge this and fix the warnings in follow-up PRs.

cdbf1 commented:
> After running `./checks/build_docs.py` locally I get
> 
> ```
> WARNING: Cannot resolve import of cirq_superstaq.ops.qubit_gates.AQTICCX in cirq_superstaq.ops
> WARNING: Cannot resolve import of cirq_superstaq.ops.qubit_gates.AQTITOFFOLI in cirq_superstaq.ops
> WARNING: Cannot resolve import of cirq_superstaq.ops.qubit_gates.CR in cirq_superstaq.ops
> WARNING: Cannot resolve import of cirq_superstaq.ops.qubit_gates.ZX in cirq_superstaq.ops
> WARNING: Cannot resolve import of cirq_superstaq.ops.AQTICCX in cirq_superstaq
> WARNING: Cannot resolve import of cirq_superstaq.ops.AQTITOFFOLI in cirq_superstaq
> WARNING: Cannot resolve import of cirq_superstaq.ops.CR in cirq_superstaq
> WARNING: Cannot resolve import of cirq_superstaq.ops.ZX in cirq_superstaq
> ```
> 
> but the docs do indeed build. I think it's fine to merge this and fix the warnings in follow-up PRs.

Turns out Sphinx didn't like the double = signs so I changed that. I also fixed (hopefully!) all the other warnings that Sphinx was generating.

It might be worth in future having the build fail if there are any Sphinx warnings.

richrines1 reviewed (commented):
adding a few nits - leaving the big picture review to @vtomole

we should also check that this works w/ the other repo before merging

cdbf1 commented on cirq-superstaq/cirq_superstaq/job.py:
```diff
@@ -189,8 +190,8 @@ def cancel(self, index: int | None = None, **kwargs: object) -> None:
-            gss.SuperstaqServerException: If unable to get the status of the job from the API or
-                cancellations were unsuccessful.
+            general_superstaq.SuperstaqServerException: If unable to get the status of the job
+               from the API orcancellations were unsuccessful.
```
This was an oversight on my part sorry. I've always used 3-spaces for indenting docstrings and thought this was a requirement, when it turns out its not.

cdbf1 commented on cirq-superstaq/cirq_superstaq/job.py:
```diff
@@ -157,13 +157,14 @@ def status(self, index: int | None = None) -> str:
+
+        Raises:
+            general_superstaq.SuperstaqServerException: If unable to get the status of the job
+               from the API.
```
I managed to fix this so we can continue using `gss`, `css`, and `qss` in docstrings. Note that the `~` which I have now included is there to make the compiled documentation look nicer by suppressing the `general_superstaq` in the return type.

vtomole commented:
@cdbf1 to confirm that this doesn't affect server docs, then this is good to go.

cdbf1 commented:
> @cdbf1 to confirm that this doesn't affect server docs, then this is good to go.

@vtomole see the minor fixes that are needed for server-superstaq here: https://github.com/Infleqtion/server-superstaq/pull/3407

vtomole reviewed (approved):
LGTM
